### DISCUSSION TITLE
Properly stop / clean / dispose

### DISCUSSION THREAD

**Discussion by lonix1** [testcontainers/testcontainers-dotnet#410] [Q&A] [answered]
I'm trying to understand how to properly dispose a container.

I use `container.StopAsync()`, and in my class' disposal code I call `container.DisposeAsync()`.

I checked the source for `DisposeAsync` and saw that it calls `CleanUpAsync` or `StopAsync`.

How do those differ? And should I be using `WithCleanUp(true)`?

**Comment by HofmeisterAn** [comment, marked as answer]
- `StopAsync` will just stop the container. All ressources remain. You can start the container in the same .NET Testcontainers run again.
- `CleanUpAsync` will remove the container resource. Be aware of [#242](https://github.com/HofmeisterAn/dotnet-testcontainers/issues/242#issuecomment-626882202). We did some further improvements, but there are edge cases where we are not able to remove the container resource yet. Anyway, if you setup a proper test, most cases will work.
- `DisposeAsync` chooses between these too implementations, it depends on whether you set `WithCleanUp` or not. By default, it is set to `true`.

It is not necessary to call `StopAsync` or `CleanUpAsync` if you are using the `await using` pattern.   `DisposeAsync` helps to reduce the cleanup code. Usually, you set `WithCleanUp` to `false` to debug and inspect your container while developing.

**Reply by lonix1** [reply]
That clears up my confusion, thank you very much! I'm sure others will benefit also from this detail.

**Comment by lonix1** [comment]
This seems to be the way to think of it, kind of:

`StartAsync()` == `docker run image`
`StopAsync()` == `docker stop container`
`CleanAsync()` == `docker rm container`
`DisposeAsync()` == `docker stop container; docker rm container`

**Comment by SeanFarrow** [comment]
This was helpful, thanks.

however I'm finding that images are not deleted.

I have both Cleanup and Remove set to true.

With is with a mssql container.

**Reply by evaapavel** [reply]
Well, I think, with Docker, there are images and there are containers. You usually do not want to delete an image (sort of a startup snapshot for your containers, you initially download it from a web repository) whereas you may want to delete a container (sort of a "used" snapshot of a virtual machine you want to get rid of after a couple of tests).
